src/plan/marksweep/global.rs

```diff
@@ -0,0 +1,183 @@
+use crate::util::alloc::malloc_allocator::HEAP_USED;
+use crate::util::heap::layout::heap_layout::Mmapper;
+use crate::util::heap::layout::heap_layout::VMMap;
+use crate::util::heap::layout::vm_layout_constants::BYTES_IN_CHUNK;
+use crate::util::heap::layout::vm_layout_constants::{HEAP_END, HEAP_START};
+use crate::util::heap::HeapMeta;
+use crate::util::malloc::free;
+use crate::util::malloc::malloc_usable_size;
+use crate::util::options::UnsafeOptionsWrapper;
+#[cfg(feature = "sanity")]
+use crate::util::sanity::sanity_checker::*;
+use crate::util::side_metadata::load_atomic;
+use crate::util::OpaquePointer;
+use crate::vm::VMBinding;
+use std::{collections::HashSet, sync::Arc};
+
+use atomic::Ordering;
+use enum_map::EnumMap;
+
+pub type SelectedPlan<VM> = MarkSweep<VM>;
+
+pub struct MarkSweep<VM: VMBinding> {
+    pub base: BasePlan<VM>,
+    pub space: MallocSpace<VM>,
+}
+
+unsafe impl<VM: VMBinding> Sync for MarkSweep<VM> {}
+
+pub const MS_CONSTRAINTS: PlanConstraints = PlanConstraints {
+    moves_objects: false,
+    gc_header_bits: 2,
+    gc_header_words: 0,
+    num_specialized_scans: 1,
+    ..PlanConstraints::default()
+};
+
+impl<VM: VMBinding> Plan for MarkSweep<VM> {
+    type VM = VM;
+
+    fn collection_required(&self, _space_full: bool, _space: &dyn Space<Self::VM>) -> bool
+    where
+        Self: Sized,
+    {
+        unsafe { HEAP_USED.load(Ordering::SeqCst) >= HEAP_SIZE }
+    }
+
+    fn gc_init(
+        &mut self,
+        heap_size: usize,
+        vm_map: &'static VMMap,
+        scheduler: &Arc<MMTkScheduler<VM>>,
+    ) {
+        unsafe {
+            HEAP_SIZE = heap_size;
+        }
+        self.base.gc_init(heap_size, vm_map, scheduler);
+    }
+
+    fn schedule_collection(&'static self, scheduler: &MMTkScheduler<VM>) {
+        self.base().set_collection_kind();
+        self.base().set_gc_status(GcStatus::GcPrepare);
+        // Stop & scan mutators (mutator scanning can happen before STW)
+        scheduler.work_buckets[WorkBucketStage::Unconstrained]
+            .add(StopMutators::<MSProcessEdges<VM>>::new());
+        // Prepare global/collectors/mutators
+        scheduler.work_buckets[WorkBucketStage::Prepare]
+            .add(Prepare::<Self, NoCopy<VM>>::new(self));
+        // Release global/collectors/mutators
+        scheduler.work_buckets[WorkBucketStage::Release]
+            .add(Release::<Self, NoCopy<VM>>::new(self));
+        // Resume mutators
+        #[cfg(feature = "sanity")]
+        scheduler.work_buckets[WorkBucketStage::Final]
+            .add(ScheduleSanityGC::<Self, NoCopy<VM>>::new());
+        scheduler.set_finalizer(Some(EndOfGC));
+    }
+
+    fn get_allocator_mapping(&self) -> &'static EnumMap<AllocationSemantics, AllocatorSelector> {
+        &*ALLOCATOR_MAPPING
+    }
+
+    fn prepare(&self, _tls: OpaquePointer) {
+        // Do nothing
+    }
+
+    fn release(&self, _tls: OpaquePointer) {
+        let mut released_chunks = HashSet::new();
```

Comment:
This is policy-specific. You should encapsulate policy-specific code to your policy `MallocSpace`. `Plan` here should simply call all the spaces for the plan to release pages.

Reply:
This is now moved to the policy.